dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java

```diff
@@ -2055,4 +2111,199 @@ public String changeJson(ProcessData processData, String oldJson) {
+        // save task definition log
+        TaskDefinitionLog taskDefinitionLog = new TaskDefinitionLog();
+        taskDefinitionLog.set(taskDefinition);
+        taskDefinitionLog.setOperator(operator.getId());
+        taskDefinitionLog.setOperateTime(now);
+        int insert = taskDefinitionLogMapper.insert(taskDefinitionLog);
+        return insert & update;
+    }
+
+    /**
+     * get resource ids
+     *
+     * @param taskDefinition taskDefinition
+     * @return resource ids
+     */
+    public String getResourceIds(TaskDefinition taskDefinition) {
+        Set<Integer> resourceIds = null;
+        // TODO modify taskDefinition.getTaskType()
+        AbstractParameters params = TaskParametersUtils.getParameters(taskDefinition.getTaskType().getDescp(), taskDefinition.getTaskParams());
+
+        if (params != null && CollectionUtils.isNotEmpty(params.getResourceFilesList())) {
+            resourceIds = params.getResourceFilesList().
+                    stream()
+                    .filter(t -> t.getId() != 0)
+                    .map(ResourceInfo::getId)
+                    .collect(Collectors.toSet());
+        }
+        if (CollectionUtils.isEmpty(resourceIds)) {
+            return StringUtils.EMPTY;
+        }
+        return StringUtils.join(resourceIds, ",");
+    }
+
+    /**
+     * @param operator
+     * @param name
+     * @param desc
+     * @param locations
+     * @param connects
+     * @param project
+     * @param processData
+     * @param processDefinition
+     * @return
+     */
+    public int saveProcessDefinition(User operator, Project project, String name, String desc, String locations,
+                                     String connects, ProcessData processData, ProcessDefinition processDefinition) {
+        List<TaskNode> taskNodeList = (processData.getTasks() == null) ? new ArrayList<>() : processData.getTasks();
+        for (TaskNode task : taskNodeList) {
+            // TODO update by code directly
+            TaskDefinition taskDefinition = taskDefinitionMapper.queryByDefinitionName(project.getCode(), task.getName());
+            updateTaskDefinition(operator, project.getCode(), task, taskDefinition);
+        }
+        createTaskAndRelation(operator, project.getName(), "", processDefinition, processData);
+        ProcessDefinitionLog processDefinitionLog = updateProcessLog(operator, processDefinition.getCode(),
+                name, processData, project, desc, locations, connects);
+        return switchVersion(processDefinition, processDefinitionLog);
+    }
+
+    /**
+     * @param operator
+     * @param processDefinitionCode
+     * @param processDefinitionName
+     * @param processData
+     * @param project
+     * @param desc
+     * @param locations
+     * @param connects
+     * @return
+     */
+    public ProcessDefinitionLog updateProcessLog(User operator, Long processDefinitionCode, String processDefinitionName,
```

Comment:
I suggest changing the method updateProcessLog to insertProcessLog，It would be more straightforward

Reply:
you a right!